File: app/src/main/java/cn/vove7/jarvis/view/custom/scmoothtextview/TextWebView.java
package cn.vove7.jarvis.view.custom.scmoothtextview;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;
import android.webkit.WebChromeClient;
import android.webkit.WebResourceRequest;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;


public class TextWebView extends WebView {

    public TextWebView(Context context) {
        super(context);
        if (isInEditMode()) return;
        initView(null);
    }

    public TextWebView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView(attrs);
    }

    public TextWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView(attrs);
    }

    @SuppressLint("SetJavaScriptEnabled")
    private void initView(@Nullable AttributeSet attrs) {
        if (isInEditMode()) return;
        if (attrs != null) {
            setBackgroundColor(0xffffff);
        }
        setWebChromeClient(new ChromeClient());
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            setWebViewClient(new WebClient());
        } else {
            setWebViewClient(new WebClientCompat());
        }
        //addJavascriptInterface();

        WebSettings settings = getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setAppCachePath(getContext().getCacheDir().getPath());
        settings.setAppCacheEnabled(true);
        settings.setSupportZoom(false);
        settings.setCacheMode(WebSettings.LOAD_NO_CACHE);
        settings.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.TEXT_AUTOSIZING);

        settings.setDefaultTextEncodingName("utf-8");
        setScrollBarStyle(View.SCROLLBARS_INSIDE_OVERLAY);
    }

    public void setSource(@NonNull String source, boolean wrap) {
        loadCode(buildHtml("<xmp style = \"" + (wrap ? "overflow: hidden" : "") + "\">\n" +
                source + "\n" +
                "    </xmp>\n" +
                source + "</xmp>"));
    }

    public void setHtml(String html) {
        loadCode(buildHtml(html));
    }

    private String baseUrl = "file:///android_asset/web/";

    private void loadCode(String page) {
        post(() -> loadDataWithBaseURL(baseUrl, page, "text/html", "utf-8", null));
    }

    private String buildHtml(String html) {
        return "<html>\n" +
                "\n" +
                "<head>\n" +
                "    <link rel=\"stylesheet\" href=\"text_style.css\">\n" +
                "</head>\n" +
                "\n" +
                "<body>\n" +
                "\n" +
                html +
                "</body>\n" +
                "\n" +
                "</html>";
    }


    private class ChromeClient extends WebChromeClient {

    }

    private void startActivity(@Nullable Uri url) {
        getContext().startActivity(new Intent(Intent.ACTION_VIEW, url));
    }

    private class WebClient extends WebViewClient {
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
            startActivity(request.getUrl());
            return true;
        }
    }

    private class WebClientCompat extends WebViewClient {
        @SuppressWarnings("deprecation")
        @Override
        public boolean shouldOverrideUrlLoading(WebView view, String url) {
            startActivity(Uri.parse(url));
            return true;
        }

    }

}

File: app/src/main/java/cn/vove7/jarvis/view/custom/scmoothtextview/TextWebView.kt
package cn.vove7.jarvis.view.custom.scmoothtextview

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.util.AttributeSet
import android.view.View
import android.webkit.*


class TextWebView : NestedWebView, TextLoader {

    private val baseUrl = "file:///android_asset/web/"

    constructor(context: Context) : super(context) {
        if (isInEditMode) return
        initView(null)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        initView(attrs)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        initView(attrs)
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun initView(attrs: AttributeSet?) {
        if (isInEditMode) return
        if (attrs != null) {
            setBackgroundColor(0xffffff)
        }
        webChromeClient = ChromeClient()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            webViewClient = WebClient()
        } else {
            webViewClient = WebClientCompat()
        }
        //addJavascriptInterface();

        val settings = settings
        settings.javaScriptEnabled = true
        settings.setAppCachePath(context.cacheDir.path)
        settings.setAppCacheEnabled(true)
        settings.setSupportZoom(false)
        settings.cacheMode = WebSettings.LOAD_NO_CACHE
        settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.TEXT_AUTOSIZING

        settings.defaultTextEncodingName = "utf-8"
        scrollBarStyle = View.SCROLLBARS_INSIDE_OVERLAY
    }

    override fun setSource(source: String, wrap: Boolean) {
        loadCode(buildHtml("<xmp style = \"" + (if (wrap) "overflow: hidden" else "") + "\">\n" +
                source + "\n" +
                "    </xmp>\n" +
                source + "</xmp>"))
    }

    override fun setHtml(html: String) {
        loadCode(buildHtml(html))
    }

    private fun loadCode(page: String) {
        post { loadDataWithBaseURL(baseUrl, page, "text/html", "utf-8", null) }
    }

    private fun buildHtml(html: String): String {
        return "<html>\n" +
                "\n" +
                "<head>\n" +
                "    <link rel=\"stylesheet\" href=\"text_style.css\">\n" +
                "</head>\n" +
                "\n" +
                "<body>\n" +
                "\n" +
                html +
                "</body>\n" +
                "\n" +
                "</html>"
    }


    private inner class ChromeClient : WebChromeClient()

    private fun startActivity(url: Uri?) {
        context.startActivity(Intent(Intent.ACTION_VIEW, url))
    }

    private inner class WebClient : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
            startActivity(request.url)
            return true
        }
    }

    private inner class WebClientCompat : WebViewClient() {
        override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
            startActivity(Uri.parse(url))
            return true
        }

    }

}

interface TextLoader {
    fun setSource(source: String, wrap: Boolean = false)
    fun setHtml(html: String)
}
